File: DanceCalc/m4d/ViewModels/DanceClass.cs
using System.Collections.Generic;
using DanceLibrary;

namespace m4d.ViewModels
{
    public class DanceMapping
    {
        public DanceMapping(string title = null, string name = null)
        {
            Title = title;
            Name = name ?? DanceObject.SeoFriendly(title);
        }

        public string Name;
        public string Title;
    }
    public class DanceClass
    {
        public string Title;
        public string Image;
        public string TopDance;

        public List<DanceMapping> Dances;
    }
}

File: m4d/Client/src/pages/home/DanceClass.ts
import "reflect-metadata";
import { jsonArrayMember, jsonMember, jsonObject } from "typedjson";

@jsonObject
export class DanceMapping {
  @jsonMember public name!: string;
  @jsonMember public title!: string;
  @jsonMember public controller!: string;
  @jsonMember public queryString?: string;

  public get link(): string {
    const query = this.queryString;
    const path = query ? `?${query}` : `/${this.name}`;
    return `/${this.controller}${path}`;
  }
}

@jsonObject
export class DanceClass {
  @jsonMember public title!: string;
  @jsonMember public fullTitle!: string;
  @jsonMember public image!: string;
  @jsonMember public topDance?: string;
  @jsonArrayMember(DanceMapping) public dances!: DanceMapping[];
}
